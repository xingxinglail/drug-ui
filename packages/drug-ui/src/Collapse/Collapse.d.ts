import * as React from 'react';
import { TransitionProps } from 'react-transition-group/Transition';
import { SimpleSpread } from '..';
interface PropsExtra {
    timeout?: number | {
        enter?: number;
        exit?: number;
    };
}
export interface CollapseProps extends SimpleSpread<TransitionProps, PropsExtra> {
    collapsedHeight?: string | number;
}
export declare const name = "Collapse";
declare const Collapse: React.FC<CollapseProps>;
export default Collapse;
