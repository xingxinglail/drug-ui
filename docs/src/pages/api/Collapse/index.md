---
filename: /packages/drug-ui/src/Collapse/Collapse.js
---

# Collapse API

<p class="description">Collapse React 组件的 API 文档。</p>

## Import

```js
import Collapse from '@drug-ui/core/Collapse';
// or
import { Collapse } from '@drug-ui/core';
```

## Props

| Name | Type | Default | Description |
|:-----|:-----|:--------|:------------|
| <span class="prop-name">children</span> | <span class="prop-type">node</span> |  | 内容节点。 |
| <span class="prop-name">collapsedHeight</span> | <span class="prop-type">number &#124; string</span> | <span class="prop-default">0px</span> | 容器折叠起来的高度。 |
| <span class="prop-name">in</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | 如果为```true```，组件将开始进入过渡。 |
| <span class="prop-name">timeout</span> | <span class="prop-type">number &#124; { enter?: number, exit?: number }</span> | <span class="prop-default">300ms</span> | 过渡时长。 |

`ref` 被转发到根元素.

## CSS

Style sheet name: `DuiCollapse`.

## Demos

- [Menu](/components/Menu/)
- [Transitions](/components/Transitions/)




