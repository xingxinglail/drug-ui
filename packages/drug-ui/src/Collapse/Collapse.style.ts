import { Style, Styles } from 'jss';
import { Theme } from '../styles';

const container: Style = {
    overflow: 'hidden',
    height: 0,
    transition: 'height .3s cubic-bezier(0.4, 0, 0.2, 1)'
};

const entered: Style = {
    overflow: 'visible',
    height: 'auto'
};

const hidden: Style = {
    overflow: 'hidden'
};

export const styles = (_: Theme): Styles => ({
    container,
    entered,
    hidden
});
