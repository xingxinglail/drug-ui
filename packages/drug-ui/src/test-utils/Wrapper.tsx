import * as React from 'react';
import { ThemeProvider } from '../index';

interface Props {
    children?: any;
}

const Wrapper: React.FC<Props> = (props) => {
    return (
        <ThemeProvider>
            { props.children }
        </ThemeProvider>
    );
};

export default Wrapper;
