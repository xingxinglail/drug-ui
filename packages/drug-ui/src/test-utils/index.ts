export { default as Wrapper } from './Wrapper';
export { getClasses } from './getClasses';

export const wait = (delay = 50) => {
    return new Promise(resolve => setTimeout(resolve, delay));
};
