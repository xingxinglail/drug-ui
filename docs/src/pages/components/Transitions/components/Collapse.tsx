import * as React from 'react';
import Collapse from '@drug-ui/core/Collapse';
import Button from '@drug-ui/core/Button';
import { createUseStyles } from '@drug-ui/styles';

const useStyles = createUseStyles({
    container: {
        display: 'flex',
        marginTop: 14,
        height: 124
    },
    box: {
        width: 100,
        height: 100,
        margin: 10,
        boxShadow: '0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14), 0px 1px 10px 0px rgba(0,0,0,0.12)'
    },
    svg: {
        width: 100,
        height: 100
    },
    polygon: {
        fill: '#fff',
        stroke: 'rgba(0, 0, 0, 0.12)',
        strokeWidth: 1
    }
}, { name: 'CollapseDemo' });

export default function CollapseDemo () {
    const classes = useStyles();

    const [checked, setChecked] = React.useState(false);

    const handleChange = () => {
        setChecked(prev => !prev);
    };

    return (
        <div>
            <Button variant="outlined" onClick={ handleChange }>Click Me</Button>
            <div className={ classes.container }>
                <Collapse in={ checked }>
                    <div className={ classes.box }>
                        <svg className={ classes.svg }>
                            <polygon points="0,100 50,00, 100,100" className={ classes.polygon } />
                        </svg>
                    </div>
                </Collapse>
                <Collapse in={ checked } collapsedHeight="40px">
                    <div className={ classes.box }>
                        <svg className={ classes.svg }>
                            <polygon points="0,100 50,00, 100,100" className={ classes.polygon } />
                        </svg>
                    </div>
                </Collapse>
            </div>
        </div>
    );
};
