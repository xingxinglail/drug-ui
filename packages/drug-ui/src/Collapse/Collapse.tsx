import * as React from 'react';
import { Transition } from 'react-transition-group';
import { TransitionProps } from 'react-transition-group/Transition';
import classnames from 'classnames';
import { styles } from './Collapse.style';
import { createUseStyles } from '../styles';
import { SimpleSpread } from '..';

interface PropsExtra {
    timeout?: number | { enter?: number, exit?: number };
}

export interface CollapseProps extends SimpleSpread<TransitionProps, PropsExtra> {
    collapsedHeight?: string | number;
}

export const name = 'Collapse';

type CollapseClassProps = 'container' | 'entered' | 'hidden' | 'wrapper' | 'innerWrapper';

const useStyles = createUseStyles<CollapseClassProps>(styles, name);

const Collapse: React.FC<CollapseProps> = React.forwardRef<HTMLDivElement, CollapseProps>((props, ref) => {
    const { className, style, collapsedHeight: collapsedHeightProp = '0px', in: inProp = false, timeout = 300, onEnter, onEntering, onEntered, onExit, onExiting, onExited, children, ...rest } = props;
    const classes = useStyles();
    const wrapperRef = React.useRef<HTMLDivElement>(null);
    const collapsedHeight = typeof collapsedHeightProp === 'number' ? `${collapsedHeightProp}px` : collapsedHeightProp;

    const handleTransitionEvent = (event: string, node: HTMLElement, isAppearing?: boolean) => {
        if (event === 'onEnter') {
            node.style.height = collapsedHeight;
            onEnter && onEnter(node, isAppearing);
        }

        if (event === 'onEntering') {
            const wrapperHeight = wrapperRef.current?.clientHeight || 0;
            node.style.transitionDuration = `${ typeof timeout === 'number' ? timeout : timeout.enter }ms`;
            node.style.height = `${ wrapperHeight }px`;
            onEntering && onEntering(node, isAppearing);
        }

        if (event === 'onEntered') {
            node.style.height = 'auto';
            onEntered && onEntered(node, isAppearing);
        }

        if (event === 'onExit') {
            const wrapperHeight = wrapperRef.current?.clientHeight || 0;
            node.style.height = `${ wrapperHeight }px`;
            onExit && onExit(node);
        }

        if (event === 'onExiting') {
            wrapperRef.current?.clientHeight;
            node.style.transitionDuration = `${ typeof timeout === 'number' ? timeout : timeout.exit }ms`;
            node.style.height = collapsedHeight;
            onExiting && onExiting(node);
        }

        if (event === 'onExited') {
            onExited && onExited(node);
        }
    };

    return (
        <Transition
            in={ inProp }
            timeout={ timeout }
            onEnter={ (node, isAppearing) => handleTransitionEvent('onEnter', node, isAppearing) }
            onEntering={ (node, isAppearing) => handleTransitionEvent('onEntering', node, isAppearing) }
            onEntered={ (node, isAppearing) => handleTransitionEvent('onEntered', node, isAppearing) }
            onExit={ (node) => handleTransitionEvent('onExit', node) }
            onExiting={ (node) => handleTransitionEvent('onExiting', node) }
            onExited={ (node) => handleTransitionEvent('onExited', node) }
            { ...rest }>
            {
                state => (
                    <div
                        ref={ ref }
                        className={
                            classnames(
                                classes.container,
                                { [classes.entered]: state === 'entered' },
                                { [classes.hidden]: state === 'exited' && !inProp && collapsedHeight === '0px' },
                                className
                            )
                        }
                        style={ { minHeight: collapsedHeight, ...style } }>
                        <div ref={ wrapperRef } className={ classes.wrapper }>
                            <div className={ classes.innerWrapper }>
                                { children }
                            </div>
                        </div>
                    </div>
                )
            }
        </Transition>
    );
});

Collapse.displayName = name;

if (process.env.NODE_ENV !== 'production') {
    // @ts-ignore
    Collapse.useStyles = useStyles;
}

export default Collapse;
