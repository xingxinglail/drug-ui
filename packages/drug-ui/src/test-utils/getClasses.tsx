import * as React from 'react';
import * as enzyme from 'enzyme';
import Wrapper from './Wrapper';

export const getClasses = (element: React.ReactElement) => {
    // @ts-ignore
    const { useStyles } = element.type;
    let classes;

    function Listener () {
        classes = useStyles();
        return null;
    }

    enzyme.mount(<Wrapper><Listener /></Wrapper>);
    return classes;
};
