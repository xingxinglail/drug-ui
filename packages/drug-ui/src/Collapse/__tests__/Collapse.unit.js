import * as renderer from 'react-test-renderer';
import * as React from 'react';
import * as enzyme from 'enzyme';
import Collapse from '..';
import { Wrapper, getClasses, wait } from '@drug-ui/core/test-utils';

const mount = enzyme.mount;

describe('<Collapse />', () => {

    let classes;

    it('render successful', () => {
        const json = renderer.create(<Wrapper><Collapse /></Wrapper>).toJSON();
        expect(json).toMatchSnapshot();
        classes = getClasses(<Collapse />);
    });

    it('默认渲染', () => {
        const wrapper = mount(<Wrapper><Collapse>Children</Collapse></Wrapper>);
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains(classes.container)).toBeTruthy();
        expect(classList.contains(classes.hidden)).toBeTruthy();
    });

    it('默认显示渲染', () => {
        const wrapper = mount(<Wrapper><Collapse in>Children</Collapse></Wrapper>);
        const classList = wrapper.getDOMNode().classList;
        expect(classList.contains(classes.container)).toBeTruthy();
        expect(classList.contains(classes.entered)).toBeTruthy();
    });

    it('collapsedHeight', (done) => {
        const onEntered = async () => {
            expect(dom.classList.contains(classes.entered)).toBeTruthy();
            expect(window.getComputedStyle(dom).getPropertyValue('height')).toEqual('auto');
            await wait();
            wrapper.setProps({ in: false });
        };

        const onExiting = async () => {
            expect(dom.classList.contains(classes.entered)).toBeFalsy();
            expect(window.getComputedStyle(dom).getPropertyValue('height')).toEqual('40px');
            done();
        };

        const wrapper = mount(
            <Wrapper>
                <Collapse collapsedHeight={ 40 } onEntered={ onEntered } onExiting={ onExiting }>
                    <div style={ { height: 200 } } />
                </Collapse>
            </Wrapper>
        );
        const dom = wrapper.getDOMNode();
        expect(window.getComputedStyle(dom).getPropertyValue('min-height')).toEqual('40px');
        wrapper.setProps({ in: true });
    });

    it('触发所有钩子', (done) => {
        const onEnter = jest.fn();
        const onEntering = jest.fn();
        const onEntered = jest.fn(async () => {
            await wait();
            wrapper.setProps({ in: false });
        });
        const onExit = jest.fn();
        const onExiting = jest.fn();
        const onExited = jest.fn(() => {
            expect(onEnter).toHaveBeenCalledBefore(onEntering);
            expect(onEntering).toHaveBeenCalledBefore(onEntered);
            expect(onEntered).toHaveBeenCalledBefore(onExit);
            expect(onExit).toHaveBeenCalledBefore(onExiting);
            expect(onExiting).toHaveBeenCalledBefore(onExited);
            done();
        });

        const wrapper = mount(
            <Wrapper>
                <Collapse
                    onEnter={ onEnter }
                    onEntering={ onEntering }
                    onEntered={ onEntered }
                    onExit={ onExit }
                    onExiting={ onExiting }
                    onExited={ onExited }>
                    <div style={ { height: 200 } } />
                </Collapse>
            </Wrapper>
        );
        const dom = wrapper.getDOMNode();
        expect(window.getComputedStyle(dom).getPropertyValue('min-height')).toEqual('0px');
        wrapper.setProps({ in: true });
    });
});
