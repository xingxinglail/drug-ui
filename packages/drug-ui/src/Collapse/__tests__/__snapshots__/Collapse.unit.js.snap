// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse /> render successful 1`] = `
<div
  className="DuiCollapse-container DuiCollapse-hidden"
  style={
    Object {
      "minHeight": "0px",
    }
  }
>
  <div
    className="DuiCollapse-wrapper"
  >
    <div
      className="DuiCollapse-innerWrapper"
    >
      <div
        className="DuiCollapse-container DuiCollapse-hidden"
        style={
          Object {
            "minHeight": "0px",
          }
        }
      >
        <div
          className="DuiCollapse-wrapper"
        >
          <div
            className="DuiCollapse-innerWrapper"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
